modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteTypeCoercion.java

```diff
@@ -281,6 +287,10 @@ protected boolean needToCast(SqlValidatorScope scope, SqlNode node, RelDataType
             return TypeUtils.customDataTypeNeedCast(typeFactory, fromType, toType);
         }
 
+        if (fromType.getSqlTypeName() == SqlTypeName.ROW) {
```

Comment:
It seems there is a bug in handling of Scalar subqueries in DML operations and this is not the way to fix it because this makes a cast from `RecordType(some_type)` to `some_type` legal, when it is obviously not the case.